### PR TITLE
win: make uv__get_osfhandle work with redirected stdio

### PR DESCRIPTION
**Pull request by poiru**
For the stdin/stdout/stderr file descriptors, we need GetStdHandle
because _get_osfhandle does not return the correct handle if the
console was attached during runtime. This may occur when e.g. a
Windows GUI application allocates a console with AllocConsole.

This is required for nodejs/node#556.

### CODE REVIEW

File: src/win/handle-inl.h [hunk 1]
```diff
@@ -164,15 +164,38 @@ INLINE static void uv_process_endgames(uv_loop_t* loop) {
 
 INLINE static HANDLE uv__get_osfhandle(int fd)
 {
+  HANDLE handle = NULL;
```
Comment (cjihrig) [inline src/win/handle-inl.h]: I think it's preferred to separate declarations from assignments.

File: src/win/handle-inl.h [hunk 2]
```diff
@@ -164,15 +164,38 @@ INLINE static void uv_process_endgames(uv_loop_t* loop) {
...
-  UV_BEGIN_DISABLE_CRT_ASSERT();
-  handle = (HANDLE) _get_osfhandle(fd);
-  UV_END_DISABLE_CRT_ASSERT();
+  /* For the stdin/stdout/stderr file descriptors, we need GetStdHandle */
+  /* because _get_osfhandle does not return the correct handle if the */
+  /* console was attached during runtime. This may occur when e.g. a Windows */
+  /* GUI application allocates a console with AllocConsole. */
+
+  switch (fd) {
+  case 0:
```
Comment (cjihrig) [inline src/win/handle-inl.h]: Can you indent the `case`s.

**Comment by poiru** [comment]
@cjihrig Done!

**Comment by saghul** [comment]
Code looks ok and a quick glance at the docs shows it does what it says on the tin. CI: https://ci.nodejs.org/job/libuv+any-pr+multi/282/

LGTM if the CI is green, but I'd like to hear @piscisaureus or @orangemocha here.

**Comment by piscisaureus** [comment]
This is not the way to go about it. 

MSVCRT keeps a table of fd-to-HANDLE mappings, and for node to work correctly this table needs to be kept up-to-date. Otherwise all kinds of things go south (e.g. `fprintf(stderr, ...)` only works if `_get_osfhandle(2)` returns the actual stderr handle.)

If you want "late" attached consoles to work properly, you need to make sure that the console output handles are attached to fd 0,1,2 instead of hacking libuv to report a fake value.

-1 on this patch

**Comment by poiru** [comment]
>If you want "late" attached consoles to work properly, you need to make sure that the console output handles are attached to fd 0,1,2 instead of hacking libuv to report a fake value.

The fileno for stdin/stdout/stderr is not necessarily 0/1/2.

>Otherwise all kinds of things go south (e.g. fprintf(stderr, ...) only works if _get_osfhandle(2) returns the actual stderr handle.)

I don't think that's true. `fprintf(stderr, ...)` will work fine as long as whoever attached the console also redirects stdio (with e.g. `freopen("CONOUT$", "w", stderr)`).

If we want to make that a requirement, we could use `_get_osfhandle(_fileno(stderr))` instead of `GetStdHandle(STD_ERROR_HANDLE)`.

**Comment by poiru** [comment]
@piscisaureus Any thoughts regarding my previous comment?

**Comment by poiru** [comment]
@piscisaureus Any thoughts regarding my previous comment?

**Comment by piscisaureus** [comment]
@poiru 
You're making _get_osfhandle() do something non-obvious, just to work around something specific node does. I don't want that to happen. 

Put the hacks and workarounds in node and see if you can convince the node.js maintainers to accept your patch.

But really, the easiest, rightful, clean way to do this is to keep "stdout" and file descriptor 2 and STD_OUTPUT_HANDLE all pointing at the same thing, and this change would be totally unnecessary.

Anyway, thanks for the effort; keep trying. Unfortunately I can't accept this patch.

**Pull request closed without merging**